### ISSUE TITLE
Needs original license.

### ISSUE THREAD

**Issue by ben-eb** [schiehll/postcss-whitespace#1]
This plugin needs to retain my original license; please update with the license in place. Thanks.

https://github.com/ben-eb/cssnano/blob/master/LICENSE-MIT#L12-L13

**Comment by schiehll** [comment]
Sure, sorry about that, I though that just keeping it as MIT was enough. I will include a license file with the content of the one that you linked. 

Actually would be awesome if you guys, that really have done all the work, could publish the cssnano-core as a plugin. I didn't ask earlier cause i needed it ASAP and I don't like to try to rush people to do open source job.

My use case is that I need to run postcss synchronous, and unfortunately cssnano is a async plugin, so if it could be synchronous as well the stand alone cssnano-core may not be necessary.

Let me know if that is something that could be relevant to the cssnano project and if I can do anything to help.

**Comment by ben-eb** [comment]
Thanks! 👍

You can run cssnano synchronously by turning off SVG compression (using this options object):

``` js
{
  svgo: false
}
```

Or, via CLI:

``` sh
cssnano --no-svgo
```

The rest of the plugins work synchronously. Just curious though, why do you need this to be sync - can't this be done with the asynchronous API?

**Comment by schiehll** [comment]
Oh nice, I've tried to find that info in the docs but didn't. Should I open an issue with the documentation label? (in the contributing guide says that's the right way to help with docs)

The need to be sync is that I'm playing with a css-in-js approach (you can find it [here](https://github.com/schiehll/suit-up)) that uses postcss to parse an ES6 template string (with CSSModules syntax `:global`, `composes`, etc), so it needs to be done in run-time because the variables in the string may change. It's not a problem to do that asynchronously, but I also need to export the parsed result as a module, and AFAIK there is no way to do that (export a module asynchronously without it being a promise and needing a `.then` in the file that's importing it) currently. So that's why I need it to run synchronously.

With this (setting `svgo: false`) it works, the only problem now is that cssnano depends on autoprefixer, that depends on caniuse-db, that have more then 3MB, which would be included in the bundle file.

I'm currently trying to do something like [inline-style-prefixer](https://github.com/rofrischmann/inline-style-prefixer) but that works with strings and as a postcss plugin, so the prefixes could be inserted only for the context that it's running and wouldn't have the 3MB downside.

All this css-in-js thing is just a POC right now, it may not be a good idea at all. Actually I would appreciate any input on this (if it's a valid try or not, etc).

TL;DR: It works really well with the svgo set to false, but I may still need the stand alone cssnano-core because of the autoprefixer dep.

**Comment by ben-eb** [comment]
I've been meaning to write an FAQ page with this information on it, as it's not exactly obvious. Feel free to open an issue. 😄 

Yeah, that makes sense; I can definitely feel your pain when it comes to client-side bundling. We're quite a way from a position to load only the necessary modules; there's a [plan for that](https://github.com/ben-eb/cssnano/issues/88) but it requires a lot of time/energy to pull off. As I am the sole maintainer of a lot of these modules now big transitions like this are much harder.

One thought I have been toying with is creating a subset of caniuse-db especially for Autoprefixer, but setting this up autonomously seems like a lot of effort for a relatively small gain. I'm sure @ai would appreciate any work towards creating something like this, my plate is already quite full.

There's a good comparison of CSS-in-JS solutions that you can find here: https://github.com/MicheleBertoli/css-in-js

**Comment by ben-eb** [comment]
I should also note that cssnano-core is not a separate PostCSS plugin as it doesn't really have a defined role; there's a lot of whitespace reduction performed via other modules which affects CSS functions (`rgba` for example), or between selectors (done via [postcss-minify-selectors](https://github.com/ben-eb/postcss-minify-selectors)); so it's not exactly complete as a plugin that should affect all insignificant whitespace. There's a lot of overlap between modules and so I've kept this as a core module with an intent to separate out code at some future point.

**Comment by schiehll** [comment]
I got it, thank you for the clarifications. 

Reading through the v4 roadmap thread I really liked the idea of making it only one plugin so the performance would be much better. The event-hooks idea sounds even better, but it would be really awesome if that was a feature of PostCSS itself, because looks like the most popular plugins for it are actually bundles of small ones, with the exception of autoprefixer i guess. So it's not really an edge use case, is the opposite of it.

In the css-in-js case, I think it's a shame that we have to create new plugins ecosystems like [jss](https://github.com/cssinjs/jss) do, when we already have more then 200 plugins that works with PostCSS, so if we could implement something like the event-hook idea in the core of the ecosystem it would be really great. I would like to help with that...i have some time and energy, but i don't know if I have the skills, or even if that's something that @ai would like to implement in the project.
